x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/tree/Node.java

```diff
@@ -161,6 +162,23 @@ protected void doCollectFirst(Predicate<? super T> predicate, List<T> matches) {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    public <R> Optional<R> collectFirstDown(Function<T, Optional<R>> fn) {
```

Comment:
This method is not needed; it's a tweaked version of `collectFirstChildren` that has several issues:

1. collect methods should return a list not a single value.
2. Optional is not used in the QL code-base (missing is null for single value, empty lists for multiple ones).
3. for condition parameter, use Predicate not Function (another reason why Optional doesn't work)
4. collectFirstDown is similar to collectFirstChildren. The former returns at first match while the latter looks at all current children in case of a match. I don't think there are any performance issues with this approach and it forces the caller to look at the issue in a more general fashion:
- do I only need for one child to match?
- what if there are multiple matches - is that acceptable or a non existing case?